Cannot set resources for sidecar tasks

### Nomad version
Nomad v0.11.1 (b43457070037800fcc8442c8ff095ff4005dab33)

### Operating system and Environment details
Debian 10.3

### Issue
When setting resource requirements for a Connect sidecar task using the `service.connect.sidecar_task` stanza, running the job results in the error:

    Error submitting job: Put "https://localhost:4646/v1/jobs": EOF

The same job file without the `sidecar_task.resources` stanza runs fine.

### Reproduction steps
Run a Docker job with some resource requirements defined under `service.connect.sidecar_task.resources`. See below for an example.

### Job file (if appropriate)
```hcl
job "echo" {
  datacenters = ["test"]
  group "echo" {
    network {
      mode = "bridge"
      port "http" {
        static = 8080
        to = 8080
      }
    }
    service {
      name = "echo"
      port = 8080
      connect {
        sidecar_service {}
        sidecar_task {
          resources {
            memory = 128
          }
        }
      }
    }
    task "echo" {
      user = 65534
      driver = "docker"
      config {
        image = "hashicorp/http-echo"
        args = [
          "-listen",
          ":8080",
          "-text",
          "hello world"
        ]
      }
      resources {
        memory = 10
      }
    }
  }
}
```

### Nomad Client logs

```
2020-05-17T15:13:36.032Z [DEBUG] http: request complete: method=PUT path=/v1/jobs duration=267.105µs
2020-05-17T15:13:36.037Z [ERROR] http: http: panic serving [::1]:56382: runtime error: invalid memory address or nil pointer dereference
goroutine 56665 [running]:
net/http.(*conn).serve.func1(0xc000c18000)
        net/http/server.go:1772 +0x139
panic(0x2a07d20, 0x55da6a0)
        runtime/panic.go:973 +0x3e3
github.com/hashicorp/nomad/command/agent.ApiResourcesToStructs(0xc0004efe50, 0x0)
        github.com/hashicorp/nomad/command/agent/job_endpoint.go:1056 +0x60
github.com/hashicorp/nomad/command/agent.apiConnectSidecarTaskToStructs(0xc001961300, 0xc000c037a0)
        github.com/hashicorp/nomad/command/agent/job_endpoint.go:1269 +0x3c
github.com/hashicorp/nomad/command/agent.ApiConsulConnectToStructs(0xc0014f1080, 0xc000c18140)
        github.com/hashicorp/nomad/command/agent/job_endpoint.go:1192 +0x58
github.com/hashicorp/nomad/command/agent.ApiServicesToStructs(0xc0014f0fa0, 0x1, 0x4, 0xc0009426c8, 0x1, 0x1)
        github.com/hashicorp/nomad/command/agent/job_endpoint.go:1177 +0x1e1
github.com/hashicorp/nomad/command/agent.ApiTgToStructsTG(0xc000c565a0, 0xc001ade000, 0xc001ade0f0)
        github.com/hashicorp/nomad/command/agent/job_endpoint.go:783 +0x1cd
github.com/hashicorp/nomad/command/agent.ApiJobToStructJob(0xc001694840, 0x2759ce0)
        github.com/hashicorp/nomad/command/agent/job_endpoint.go:768 +0x4a7
github.com/hashicorp/nomad/command/agent.(*HTTPServer).jobUpdate(0xc000429090, 0x37b77c0, 0xc000687260, 0xc001d3fe00, 0x0, 0x0, 0x0, 0xc002027970, 0x8, 0xc002027930)  
        github.com/hashicorp/nomad/command/agent/job_endpoint.go:408 +0x1d8
github.com/hashicorp/nomad/command/agent.(*HTTPServer).JobsRequest(0xc000429090, 0x37b77c0, 0xc000687260, 0xc001d3fe00, 0x4d6f96, 0x5ec154a0, 0x1f0d59f, 0x368ac3774721)
        github.com/hashicorp/nomad/command/agent/job_endpoint.go:22 +0x84
github.com/hashicorp/nomad/command/agent.(*HTTPServer).wrap.func1(0x37b77c0, 0xc000687260, 0xc001d3fe00)
        github.com/hashicorp/nomad/command/agent/http.go:445 +0x176
net/http.HandlerFunc.ServeHTTP(0xc0004f6bc0, 0x37b77c0, 0xc000687260, 0xc001d3fe00)
        net/http/server.go:2012 +0x44
net/http.(*ServeMux).ServeHTTP(0xc0008ef4c0, 0x37b77c0, 0xc000687260, 0xc001d3fe00)
        net/http/server.go:2387 +0x1a5
github.com/hashicorp/nomad/vendor/github.com/NYTimes/gziphandler.GzipHandlerWithOpts.func1.1(0x37bb840, 0xc00159e0e0, 0xc001d3fe00)
        github.com/hashicorp/nomad/vendor/github.com/NYTimes/gziphandler/gzip.go:269 +0x1e6
net/http.HandlerFunc.ServeHTTP(0xc00092ec60, 0x37bb840, 0xc00159e0e0, 0xc001d3fe00)
        net/http/server.go:2012 +0x44
net/http.serverHandler.ServeHTTP(0xc0004d2620, 0x37bb840, 0xc00159e0e0, 0xc001d3fe00)
        net/http/server.go:2807 +0xa3
net/http.(*conn).serve(0xc000c18000, 0x37d28c0, 0xc001c03680)
        net/http/server.go:1895 +0x86c
created by net/http.(*Server).Serve
        net/http/server.go:2933 +0x35c
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,10 @@ IMPROVEMENTS:
 
  * build: Updated to Go 1.14.3 [[GH-7431](https://github.com/hashicorp/nomad/issues/7970)]
 
+BUG FIXES:
+
+ * api: Fixed a bug where setting connect sidecar task resources could fail [[GH-7993](https://github.com/hashicorp/nomad/issues/7993)]
+
 ## 0.11.2 (May 14, 2020)
 
 FEATURES:

--- a/api/services.go
+++ b/api/services.go
@@ -128,6 +128,8 @@ func (s *Service) Canonicalize(t *Task, tg *TaskGroup, job *Job) {
 		s.AddressMode = "auto"
 	}
 
+	s.Connect.Canonicalize()
+
 	// Canonicalize CheckRestart on Checks and merge Service.CheckRestart
 	// into each check.
 	for i, check := range s.Checks {
@@ -143,12 +145,33 @@ type ConsulConnect struct {
 	SidecarTask    *SidecarTask          `mapstructure:"sidecar_task"`
 }
 
+func (cc *ConsulConnect) Canonicalize() {
+	if cc == nil {
+		return
+	}
+
+	cc.SidecarService.Canonicalize()
+	cc.SidecarTask.Canonicalize()
+}
+
 // ConsulSidecarService represents a Consul Connect SidecarService jobspec
 // stanza.
 type ConsulSidecarService struct {
 	Tags  []string
 	Port  string
 	Proxy *ConsulProxy
+}
+
+func (css *ConsulSidecarService) Canonicalize() {
+	if css == nil {
+		return
+	}
+
+	if len(css.Tags) == 0 {
+		css.Tags = nil
+	}
+
+	css.Proxy.Canonicalize()
 }
 
 // SidecarTask represents a subset of Task fields that can be set to override
@@ -167,6 +190,44 @@ type SidecarTask struct {
 	KillSignal    string         `mapstructure:"kill_signal"`
 }
 
+func (st *SidecarTask) Canonicalize() {
+	if st == nil {
+		return
+	}
+
+	if len(st.Config) == 0 {
+		st.Config = nil
+	}
+
+	if len(st.Env) == 0 {
+		st.Env = nil
+	}
+
+	if st.Resources == nil {
+		st.Resources = DefaultResources()
+	} else {
+		st.Resources.Canonicalize()
+	}
+
+	if st.LogConfig == nil {
+		st.LogConfig = DefaultLogConfig()
+	} else {
+		st.LogConfig.Canonicalize()
+	}
+
+	if len(st.Meta) == 0 {
+		st.Meta = nil
+	}
+
+	if st.KillTimeout == nil {
+		st.KillTimeout = timeToPtr(5 * time.Second)
+	}
+
+	if st.ShutdownDelay == nil {
+		st.ShutdownDelay = timeToPtr(0)
+	}
+}
+
 // ConsulProxy represents a Consul Connect sidecar proxy jobspec stanza.
 type ConsulProxy struct {
 	LocalServiceAddress string              `mapstructure:"local_service_address"`
@@ -174,6 +235,22 @@ type ConsulProxy struct {
 	ExposeConfig        *ConsulExposeConfig `mapstructure:"expose"`
 	Upstreams           []*ConsulUpstream
 	Config              map[string]interface{}
+}
+
+func (cp *ConsulProxy) Canonicalize() {
+	if cp == nil {
+		return
+	}
+
+	cp.ExposeConfig.Canonicalize()
+
+	if len(cp.Upstreams) == 0 {
+		cp.Upstreams = nil
+	}
+
+	if len(cp.Config) == 0 {
+		cp.Config = nil
+	}
 }
 
 // ConsulUpstream represents a Consul Connect upstream jobspec stanza.
@@ -184,6 +261,16 @@ type ConsulUpstream struct {
 
 type ConsulExposeConfig struct {
 	Path []*ConsulExposePath `mapstructure:"path"`
+}
+
+func (cec *ConsulExposeConfig) Canonicalize() {
+	if cec == nil {
+		return
+	}
+
+	if len(cec.Path) == 0 {
+		cec.Path = nil
+	}
 }
 
 type ConsulExposePath struct {

--- a/api/services_test.go
+++ b/api/services_test.go
@@ -4,7 +4,6 @@ import (
 	"testing"
 	"time"
 
-	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
@@ -45,22 +44,110 @@ func TestService_CheckRestart(t *testing.T) {
 	}
 
 	service.Canonicalize(task, tg, job)
-	assert.Equal(t, service.Checks[0].CheckRestart.Limit, 22)
-	assert.Equal(t, *service.Checks[0].CheckRestart.Grace, 22*time.Second)
-	assert.True(t, service.Checks[0].CheckRestart.IgnoreWarnings)
+	require.Equal(t, service.Checks[0].CheckRestart.Limit, 22)
+	require.Equal(t, *service.Checks[0].CheckRestart.Grace, 22*time.Second)
+	require.True(t, service.Checks[0].CheckRestart.IgnoreWarnings)
 
-	assert.Equal(t, service.Checks[1].CheckRestart.Limit, 33)
-	assert.Equal(t, *service.Checks[1].CheckRestart.Grace, 33*time.Second)
-	assert.True(t, service.Checks[1].CheckRestart.IgnoreWarnings)
+	require.Equal(t, service.Checks[1].CheckRestart.Limit, 33)
+	require.Equal(t, *service.Checks[1].CheckRestart.Grace, 33*time.Second)
+	require.True(t, service.Checks[1].CheckRestart.IgnoreWarnings)
 
-	assert.Equal(t, service.Checks[2].CheckRestart.Limit, 11)
-	assert.Equal(t, *service.Checks[2].CheckRestart.Grace, 11*time.Second)
-	assert.True(t, service.Checks[2].CheckRestart.IgnoreWarnings)
+	require.Equal(t, service.Checks[2].CheckRestart.Limit, 11)
+	require.Equal(t, *service.Checks[2].CheckRestart.Grace, 11*time.Second)
+	require.True(t, service.Checks[2].CheckRestart.IgnoreWarnings)
 }
 
-// TestService_Connect asserts Service.Connect settings are properly
-// inherited by Checks.
-func TestService_Connect(t *testing.T) {
+func TestService_Connect_Canonicalize(t *testing.T) {
+	t.Parallel()
+
+	t.Run("nil connect", func(t *testing.T) {
+		cc := (*ConsulConnect)(nil)
+		cc.Canonicalize()
+		require.Nil(t, cc)
+	})
+
+	t.Run("empty connect", func(t *testing.T) {
+		cc := new(ConsulConnect)
+		cc.Canonicalize()
+		require.False(t, cc.Native)
+		require.Nil(t, cc.SidecarService)
+		require.Nil(t, cc.SidecarTask)
+	})
+}
+
+func TestService_Connect_ConsulSidecarService_Canonicalize(t *testing.T) {
+	t.Parallel()
+
+	t.Run("nil sidecar_service", func(t *testing.T) {
+		css := (*ConsulSidecarService)(nil)
+		css.Canonicalize()
+		require.Nil(t, css)
+	})
+
+	t.Run("empty sidecar_service", func(t *testing.T) {
+		css := new(ConsulSidecarService)
+		css.Canonicalize()
+		require.Empty(t, css.Tags)
+		require.Nil(t, css.Proxy)
+	})
+
+	t.Run("non-empty sidecar_service", func(t *testing.T) {
+		css := &ConsulSidecarService{
+			Tags: make([]string, 0),
+			Port: "port",
+			Proxy: &ConsulProxy{
+				LocalServiceAddress: "lsa",
+				LocalServicePort:    80,
+			},
+		}
+		css.Canonicalize()
+		require.Equal(t, &ConsulSidecarService{
+			Tags: nil,
+			Port: "port",
+			Proxy: &ConsulProxy{
+				LocalServiceAddress: "lsa",
+				LocalServicePort:    80},
+		}, css)
+	})
+}
+
+func TestService_Connect_ConsulProxy_Canonicalize(t *testing.T) {
+	t.Parallel()
+
+	t.Run("nil proxy", func(t *testing.T) {
+		cp := (*ConsulProxy)(nil)
+		cp.Canonicalize()
+		require.Nil(t, cp)
+	})
+
+	t.Run("empty proxy", func(t *testing.T) {
+		cp := new(ConsulProxy)
+		cp.Canonicalize()
+		require.Empty(t, cp.LocalServiceAddress)
+		require.Zero(t, cp.LocalServicePort)
+		require.Nil(t, cp.ExposeConfig)
+		require.Nil(t, cp.Upstreams)
+		require.Empty(t, cp.Config)
+	})
+
+	t.Run("non empty proxy", func(t *testing.T) {
+		cp := &ConsulProxy{
+			LocalServiceAddress: "127.0.0.1",
+			LocalServicePort:    80,
+			ExposeConfig:        new(ConsulExposeConfig),
+			Upstreams:           make([]*ConsulUpstream, 0),
+			Config:              make(map[string]interface{}),
+		}
+		cp.Canonicalize()
+		require.Equal(t, "127.0.0.1", cp.LocalServiceAddress)
+		require.Equal(t, 80, cp.LocalServicePort)
+		require.Equal(t, &ConsulExposeConfig{}, cp.ExposeConfig)
+		require.Nil(t, cp.Upstreams)
+		require.Nil(t, cp.Config)
+	})
+}
+
+func TestService_Connect_proxy_settings(t *testing.T) {
 	t.Parallel()
 
 	job := &Job{Name: stringToPtr("job")}
@@ -84,9 +171,9 @@ func TestService_Connect(t *testing.T) {
 
 	service.Canonicalize(task, tg, job)
 	proxy := service.Connect.SidecarService.Proxy
-	assert.Equal(t, proxy.Upstreams[0].LocalBindPort, 80)
-	assert.Equal(t, proxy.Upstreams[0].DestinationName, "upstream")
-	assert.Equal(t, proxy.LocalServicePort, 8000)
+	require.Equal(t, proxy.Upstreams[0].LocalBindPort, 80)
+	require.Equal(t, proxy.Upstreams[0].DestinationName, "upstream")
+	require.Equal(t, proxy.LocalServicePort, 8000)
 }
 
 func TestService_Tags(t *testing.T) {
@@ -107,4 +194,36 @@ func TestService_Tags(t *testing.T) {
 	r.True(service.EnableTagOverride)
 	r.Equal([]string{"a", "b"}, service.Tags)
 	r.Equal([]string{"c", "d"}, service.CanaryTags)
+}
+
+func TestService_Connect_SidecarTask_Canonicalize(t *testing.T) {
+	t.Parallel()
+
+	t.Run("nil sidecar_task", func(t *testing.T) {
+		st := (*SidecarTask)(nil)
+		st.Canonicalize()
+		require.Nil(t, st)
+	})
+
+	t.Run("empty sidecar_task", func(t *testing.T) {
+		st := new(SidecarTask)
+		st.Canonicalize()
+		require.Nil(t, st.Config)
+		require.Nil(t, st.Env)
+		require.Equal(t, DefaultResources(), st.Resources)
+		require.Equal(t, DefaultLogConfig(), st.LogConfig)
+		require.Nil(t, st.Meta)
+		require.Equal(t, 5*time.Second, *st.KillTimeout)
+		require.Equal(t, 0*time.Second, *st.ShutdownDelay)
+	})
+
+	t.Run("non empty sidecar_task resources", func(t *testing.T) {
+		exp := DefaultResources()
+		exp.MemoryMB = intToPtr(333)
+		st := &SidecarTask{
+			Resources: &Resources{MemoryMB: intToPtr(333)},
+		}
+		st.Canonicalize()
+		require.Equal(t, exp, st.Resources)
+	})
 }

--- a/api/tasks.go
+++ b/api/tasks.go
@@ -444,6 +444,7 @@ func (g *TaskGroup) Canonicalize(job *Job) {
 	if g.Name == nil {
 		g.Name = stringToPtr("")
 	}
+
 	if g.Count == nil {
 		if g.Scaling != nil && g.Scaling.Min != nil {
 			g.Count = intToPtr(int(*g.Scaling.Min))
@@ -676,6 +677,7 @@ func (t *Task) Canonicalize(tg *TaskGroup, job *Job) {
 		t.Resources = &Resources{}
 	}
 	t.Resources.Canonicalize()
+
 	if t.KillTimeout == nil {
 		t.KillTimeout = timeToPtr(5 * time.Second)
 	}

